### ijob -c1 -p standard -A berglandlab
### module load intel/18.0 intelmpi/18.0 R/3.6.0; R

#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

#BiocManager::install("SeqArray")

### libraries
  library(data.table)
  #library(cowplot)
  library(foreach)
  library(SeqArray)

### load QTLSeqR output
  #load("/mnt/sammas_storage/bergland-lab/alan/harp_summarized_play.Rdata")
  load(file="/nv/vol186/bergland-lab/alan/peaks.Rdata")
  setnames(peaks, "CHROM", "chr")

### load SuperClone & SNP filter file
  #sc <- fread("/project/berglandlab/Karen/MappingDec2019/Superclones201617182019pulexonlyD82016problematic_20200122")
  sc <- fread("/project/berglandlab/Karen/MappingDec2019/CloneInfoFilePulexandObtusa_withmedrd_20200207")
  snps2use <- fread("/project/berglandlab/Karen/MappingDec2019/snpsvarpulexpresentinhalf_table_20200207")
  setnames(snps2use, "variant.ids", "id")
  snps2use[,use:=T]

  setkey(snps2use, chr, pos)

### open GDS
  genofile <- seqOpen("/project/berglandlab/Karen/MappingDec2019/MapDec19PulexandObtusaandPulicaria_filtsnps10bpindels_snps_filter_pass_lowGQmiss_ann.seq.gds")
  snp.dt <- data.table(chr=seqGetData(genofile, "chromosome"), pos=seqGetData(genofile, "position"), id=seqGetData(genofile, "variant.id"), key="id")
  setkey(snp.dt, chr, pos)

  snp.dt <- merge(snp.dt, snps2use)

  snp.dt[,use.chr:=F]
  snp.dt[chr%in%snp.dt[,.N,chr][N>1000]$chr, use.chr:=T]


### load haplotype prior file of A & B; this is generated by `DaphniaPulex20162017Sequencing/AlanAnalysis/AB_phasing/whatshap.sh`
  dat <- fread("/nv/vol186/bergland-lab/alan/testTrio.consensus.header.phase.csv")

  setnames(dat, c(1,2), c("chr", "pos"))

  dat <- dat[!(A=="1/1" & B=="1/1")][!(A=="0/0" & B=="0/0")]

  dat.p <- dat[,list(ref=REF, alt=ALT,
                      A1=as.numeric(substr(A, 0, 1)),
                      A2=as.numeric(substr(A, 3, 3)),
                      B1=as.numeric(substr(B, 0, 1)),
                      B2=as.numeric(substr(B, 3, 3))),
                  list(chr, pos)]

  setkey(dat.p, chr, pos)
  setkey(snp.dt, chr, pos)
  dat.p <- merge(dat.p, snp.dt)

  dat.p[,A.geno:=A1+A2]
  dat.p[,B.geno:=B1+B2]

### Identify F1s
  seqResetFilter(genofile)
  seqSetFilterPos(genofile, chr=dat.p$chr, pos=dat.p$pos)

  genomat <- seqGetData(genofile, "$dosage")

  het.count <- foreach(i=1:dim(genomat)[1], .combine="rbind")%do%{
    #i<-100
    print(paste(i, dim(genomat)[1], sep=" / "))

    tmp <- dat.p
    tmp[,geno:=genomat[i,]]
    nLoci <- dim(tmp[!is.na(A.geno) & !is.na(B.geno) & !is.na(geno)])[1]
    tmp.ag <- tmp[!is.na(A.geno) & !is.na(B.geno),list(nRR=sum(geno==2, na.rm=T),
                                                      nRA=sum(geno==1, na.rm=T),
                                                      nAA=sum(geno==0, na.rm=T),
                                                      n=sum(!is.na(geno))),
                                                  list(A.geno=abs(A.geno-2), B.geno=abs(B.geno-2))]
    tmp.ag[,fRR:=nRR/n]
    tmp.ag[,fRA:=nRA/n]
    tmp.ag[,fAA:=nAA/n]

    tmp.ag[,sample.id:=seqGetData(genofile, "sample.id")[i]]
    tmp.ag
  }
  f1.set <- het.count[n>1000][A.geno==0 & B.geno==2 & fRA>.9]$sample.id

### get HWE stats for QTL of interest in F1s
  getGeno <- function(qtl.i, f1s=f1.set) {
    #qtl.i<-7
    seqResetFilter(genofile)

    region=dat.p[chr==peaks$chr[qtl.i]][pos>=peaks$start[qtl.i]][pos<=peaks$end[qtl.i]]
    region[,A1B1:=A1+B1]
    region.ag <- region[,list(exp=c(A1+B1, A1+B2, A2+B1, A2+B2), geno=c("A1_B1", "A1_B2", "A2_B1", "A2_B2")), list(chr, pos, id=id.x)]

    seqSetFilter(genofile, sample.id=f1.set, variant.id=region$id.x)

    genomat <- seqGetData(genofile, "$dosage")
    dimnames(genomat) <- list(f1.set, region$id.x)

    genomat.dt <- as.data.table(as.data.frame.table(genomat))
    setnames(genomat.dt, names(genomat.dt), c("sample.id", "id", "geno"))
    genomat.dt[,id:=as.numeric(as.character(id))]
    genomat.dt[,sample.id:=(as.character(sample.id))]

    genomat.dt <- merge(genomat.dt, region.ag, by="id", allow.cartesian=T)

    genomat.dt[,qtl.i:=qtl.i]

    return(genomat.dt)
  }

  genos.qtl <- foreach(i=1:dim(peaks)[1])%do%getGeno(qtl.i=i)
  save(genos.qtl, file="/nv/vol186/bergland-lab/alan/qtlGeno.Rdata")

#### load
  # cp /project/berglandlab/Karen/MappingDec2019/CloneInfoFilePulexandObtusa_withmedrd_20200207 /nv/vol186/bergland-lab/alan/.

  library(data.table)
  library(ggplot2)

  load("/mnt/sammas_storage/bergland-lab/alan/qtlGeno.Rdata")
  sc <- fread("/mnt/sammas_storage/bergland-lab/alan/CloneInfoFilePulexandObtusa_withmedrd_20200207")
  setnames(sc, "clone", "sample.id")

  sc[,SC.uniq := SC]
  sc[SC=="OO", SC.uniq:=paste("OO", c(1:sum(SC=="OO")))]

  ### one type of plot
    ggplot(data=genos.qtl[qtl.i==12], aes(x=as.factor(pos), y=sample.id, fill=geno.x)) + geom_tile()

  ### most likely phase, plot
    genos.qtl.ag <- genos.qtl[,list(hmd=sum(geno.x - exp, na.rm=T), geno.x=mean(geno.x, na.rm=T)), list(sample.id, qtl.i, geno.y, pos)]
    genos.qtl.ag.ag <- genos.qtl.ag[,list(geno=geno.y[which.min(abs(hmd))], hmd=min(abs(hmd)), geno.obs=mean(geno.x, na.rm=T),
                                          ),
                                     list(sample.id, qtl.i, pos)]

    setkey(genos.qtl.ag.ag, sample.id)
    setkey(sc, sample.id)

    genos.qtl.ag.ag <- merge(genos.qtl.ag.ag, sc)

    ranked <- genos.qtl.ag.ag[,list(.N), (SC)]
    ranked[,ord:=rank(N, ties="first")]

    genos.qtl.ag.ag <- merge(genos.qtl.ag.ag, ranked, by="SC")

    genos.qtl.ag.ag[,ordf:=factor(ord, levels=sort(unique(genos.qtl.ag.ag$ord)))]

    ggplot(data=(genos.qtl.ag.ag[qtl.i==11][order(SCnum)]), aes(x=as.factor(pos), y=interaction(sample.id, ordf), fill=as.factor(geno))) +
    geom_tile() + facet_wrap(~ordf)


    ggplot(data=genos.qtl.ag[qtl.i==11][sample.id=="April17_2018_D8_Male1"], aes(x=as.factor(pos), y=geno.y, fill=as.factor(hmd))) + geom_tile()


  genos.qtl.ag <- genos.qtl[,list(hmd=sum(geno.x - exp, na.rm=T)), list(sample.id, qtl.i, geno.y)]
  genos.qtl.ag.ag <- genos.qtl.ag[,list(geno=geno.y[which.min(abs(hmd))], hmd=min(abs(hmd))), list(sample.id, qtl.i)]

  ggplot(data=genos.qtl.ag.ag[qtl.i==12], aes(x=as.factor(pos), y=sample.id, fill=geno)) + geom_tile()



  genos.qtl.ag.ag[,A.geno:=tstrsplit(geno, "_")[[1]]]
  genos.qtl.ag.ag[,B.geno:=tstrsplit(geno, "_")[[2]]]

  genos.qtl.ag.ag[,A.geno:=factor(A.geno, levels=c("A1", "A2"))]
  genos.qtl.ag.ag[,B.geno:=factor(B.geno, levels=c("B1", "B2"))]

  fewt <- genos.qtl.ag.ag[,list(p=fisher.test(table(A.geno, B.geno))$p.value, fet=fisher.test(table(A.geno, B.geno))$estimate),
                            list(qtl.i)]

chisq.test(table(genos.qtl.ag.ag[qtl.i==12]$A.geno,
                 genos.qtl.ag.ag[qtl.i==12]$B.geno))







(this comes from `DaphniaPulex20162017Sequencing/AlanAnalysis/AB_phasing/prepare_vcf.R`)





    ### 1. identify fixed difference between A&B
      ab.fd <- foreach(sc.i=c("A", "B"), .combine="cbind")%do%{
        seqResetFilter(genofile)
        seqSetFilter(genofile, sample.id=sc[SC==sc.i]$clone, variant.id=snp.dt$id.x)

        data.table(af=seqAlleleFreq(genofile))
      }
      setnames(ab.fd, c(1,2), c("af.A", "af.B"))
      ab.fd <- cbind(ab.fd, snp.dt)
      ab.fd[!is.na(af.A),A.geno := unlist(sapply(ab.fd[!is.na(af.A)]$af.A, function(x) c(0,1,2)[which.min(abs(x-c(0,.5,1)))]))]
      ab.fd[!is.na(af.B),B.geno := unlist(sapply(ab.fd[!is.na(af.B)]$af.B, function(x) c(0,1,2)[which.min(abs(x-c(0,.5,1)))]))]


    ### 2. Identify F1s
      seqResetFilter(genofile)
      seqSetFilter(genofile, variant.id=snp.dt$id.x)

      genomat <- seqGetData(genofile, "$dosage")

      het.count <- foreach(i=1:dim(genomat)[1], .combine="rbind")%do%{
        print(paste(i, dim(genomat)[1], sep=" / "))

        tmp <- ab.fd
        tmp[,geno:=genomat[i,]]
        nLoci <- dim(tmp[!is.na(A.geno) & !is.na(B.geno) & !is.na(geno)])[1]
        tmp.ag <- tmp[!is.na(A.geno) & !is.na(B.geno),list(nRR=sum(geno==2, na.rm=T),
                                                          nRA=sum(geno==1, na.rm=T),
                                                          nAA=sum(geno==0, na.rm=T),
                                                          n=sum(!is.na(geno))),
                                                      list(A.geno=abs(A.geno-2), B.geno=abs(B.geno-2))]
        tmp.ag[,fRR:=nRR/n]
        tmp.ag[,fRA:=nRA/n]
        tmp.ag[,fAA:=nAA/n]

        tmp.ag[,sample.id:=seqGetData(genofile, "sample.id")[i]]
        tmp.ag
      }

      f1.set <- het.count[n>1000][A.geno==0 & B.geno==2 & fRA>.9]$sample.id
